src/goto-programs/string_refine_preprocess.cpp

```diff
@@ -0,0 +1,1343 @@
+  make_string_function(
+    goto_program, i_it, function_name, signature, true, false);
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::build_function_application
+
+  Inputs: a function name, a type, a location and a vector of arguments
+
+ Outputs: a function application expression
+
+ Purpose: declare a function and construct an function application expression
+          with the given function name, type, location and arguments
+
+\*******************************************************************/
+
+function_application_exprt
+  string_refine_preprocesst::build_function_application(
+    const irep_idt &function_name,
+    const typet &type,
+    const source_locationt &location,
+    const exprt::operandst &arguments)
+{
+  declare_function(function_name, type);
+  function_application_exprt function_app;
+  function_app.type()=type;
+  function_app.add_source_location()=location;
+  function_app.function()=symbol_exprt(function_name);
+  for(auto arg : arguments)
+    function_app.arguments().push_back(arg);
+
+  return function_app;
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::make_to_char_array_function
+
+  Inputs: a goto program and a position in that goto program
+
+ Purpose: at the given position replace `return_tmp0=s.toCharArray()` with:
+          > return_tmp0->data=&((s->data)[0])
+          > return_tmp0->length=s->length
+
+\*******************************************************************/
+
+void string_refine_preprocesst::make_to_char_array_function(
+  goto_programt &goto_program, goto_programt::targett &i_it)
+{
+  const code_function_callt &function_call=to_code_function_call(i_it->code);
+
+  assert(function_call.arguments().size()>=1);
+  const exprt &string_argument=function_call.arguments()[0];
+  assert(refined_string_typet::is_java_string_pointer_type(
+           string_argument.type()));
+
+  dereference_exprt deref(string_argument, string_argument.type().subtype());
+  typet length_type, data_type;
+  get_data_and_length_type_of_string(deref, data_type, length_type);
+  member_exprt rhs_data(deref, "data", data_type);
+  member_exprt rhs_length(deref, "length", length_type);
+  dereference_exprt rhs_array(rhs_data, data_type.subtype());
+  exprt first_index=from_integer(0, java_int_type());
+  index_exprt first_element(rhs_array, first_index, java_char_type());
+  address_of_exprt rhs_pointer(first_element);
+  typet deref_type=function_call.lhs().type().subtype();
+  dereference_exprt deref_lhs(function_call.lhs(), deref_type);
+  member_exprt lhs_data(deref_lhs, "data", data_type);
+  member_exprt lhs_length(deref_lhs, "length", length_type);
+  std::list<code_assignt> assignments;
+  assignments.emplace_back(lhs_data, rhs_pointer);
+  assignments.emplace_back(lhs_length, rhs_length);
+  insert_assignments(goto_program, i_it, assignments);
```

Comment:
I find the above really hard to read for it does not have an obvious structure, but I trust it does the right thing...